addon/inject.ts

```diff
@@ -4,32 +4,91 @@ import { getOwner } from '@ember/application';
 import { Registry } from './index';
 import Provider from './provider';
 
-const HAVE_MODIFIED = new WeakSet<EmberObject>();
-const PROVIDERS = new WeakMap<EmberObject, Array<Provider>>();
+const INJECTED_PROVIDERS = new WeakMap<EmberObject, Provider[]>();
 
-export default function inject<K extends keyof Registry>(identifier: K) {
+interface IProviderKlass extends Function {
+  create: () => Provider;
+}
+
+interface IEmberObjectWithParentView extends EmberObject {
+  parentView?: IEmberObjectWithParentView;
+}
+
+function findParentProvider(
+  component: IEmberObjectWithParentView,
+  ProviderKlass: IProviderKlass
+): Provider | undefined {
+  if (component.parentView) {
```

Comment:
(Random other thought — does `parentView` come with [Custom Components](https://github.com/emberjs/rfcs/blob/master/text/0213-custom-components.md)...haven't tried, but I'd be curious)

Reply:
Not sure. I played around with that API a little bit, but it didn't do what I wanted it to (which was coming up with a nicer way to render React components in Ember than my current hack).

If we went that direction, it would just require that anyone injecting a Provider extends a specific base class that we provide, which wouldn't necessarily be bad, but _would_ mean that this addon only supports the newest version of Ember (and we're quite behind in my app at work)